Name individual frameworks, but call out subroutines (#85)
cfrg/draft-irtf-cfrg-generic-hybrid-kems: draft-irtf-cfrg-hybrid-kems.md

Reviewer: ```suggestion
This document defines generic constructions for hybrid key encapsulation
mechanisms (KEMs) based on combining a post-quantum (PQ) KEM with a traditional
```

The security requirements for KDFs used with the frameworks in this document are
laid out in {{security-kdfs}}.

> # Hybrid KEMs {#frameworks}

Reviewer: We aren't defining concrete instances...

> # Hybrid KEMs {#frameworks}

Reviewer: I don't love this but what about: 

```suggestion
# Hybrid KEMs (Generic) {#frameworks}
```

In this section, we define four frameworks for building hybrid KEMs.  These
frameworks are based on a common set of subroutines for things like key
generation and computing a final shared secret.

The four frameworks vary along two axes:

1. Whether traditional component is a nominal group or a KEM
2. Whether to rely on the C2PRI property for the post-quantum component

> The choice of which framwork to use when building a hybrid KEM will depend on

Reviewer: ```suggestion
The choice of which framework to use when building a hybrid KEM will depend on
```

the application's needs along these two axes.

| Name    | PQ C2PRI? | T component   |
|:========|:==========|:==============|
| UG      | No        | Nominal group |
| UK      | No        | KEM           |
| CG      | Yes       | Nominal group |
| CK      | Yes       | KEM           |
{: #variants title="Hybrid KEM frameworks" }

> These frameworks create a hybrid KEM `KEM_H` based on the following constituent

Author: ```suggestion
Instantiating one of these frameworks creates a hybrid KEM `KEM_H` based on the following constituent
```

components:

* A traditional component that is either a nominal group or a KEM:
    * `Group_T`: A nominal group
    * `KEM_T`: A traditional KEM
* `KEM_PQ`: A post-quantum KEM
* `PRG`: A PRG producing byte strings of length `KEM_PQ.Nseed +
  Comp_T.Nseed` (`PRG.Nout == KEM_PQ.Nseed + Comp_T.Nseed`)
* `KDF`: A KDF producing byte strings of length `KEM_H.Nss` (`KDF.Nout
  == KEM_H.Nss`)
* `Label` - A byte string used to label the specific combination of the above

>   constituents being used, as well as which framework is being instantiated.

Author: ```suggestion
  components being used, as well as which framework is being instantiated.
```

  This value should be registered in the Hybrid KEM

>   Labels IANA registry to avoid conflict with other instantiations (see

Author: GitHub seems to have broken the link, but I'm assuming it's the comment about "secretly registering concrete hybrid KEM constructions"?  Not sure what the concern is there.  The registry is First Come First Served, so (a) it's non-rivalrous, and (b) it's not hard to register something (just email IANA).  The only way to get around this would be to remove the information from the IANA registry about what the label indicates, which seems like it removes most of the utility from the registry.  In any case, this issue seems separate from this PR.

  {{iana-considerations}}.

`KEM_PQ`, `Group_T`, `PRG`, and `KDF` MUST meet the interfaces
described in {{cryptographic-deps}} and MUST meet the security requirements
described in {{hybrid-ind-cca}}.

The constants for public values are derived from the concatenation of
encapsulation keys and ciphertexts:

~~~
KEM_H.Nek = KEM_PQ.Nek + (KEM_T.Nek or Group_T.Nelem)
KEM_H.Nct = KEM_PQ.Nct + (KEM_T.Nct or Group_T.Nelem)
~~~

The `Nseed` and `Nss` constants should reflect the overall security level of
the combined KEM, with the following recommended values:

~~~
KEM_H.Nseed = max(KEM_PQ.Nseed, (KEM_T.Nseed or Group_T.Nseed))
KEM_H.Nss = min(KEM_PQ.Nss, (KEM_T.Nss or Group_T.Nss))
~~~

Since we use the seed as the decapsulation key, `Ndk = Nseed`.  For legacy cases
where it is not possible to derive per-component decapsulation keys from a
common seed, see {{key-generation}}.

## Subroutines

> The four hybrid KEM framework share a substantial amount of structure, which we

Reviewer: ```suggestion
The four hybrid KEM frameworks share a substantial amount of structure, which we
```

capture in a set of subroutines.

### Using a Nominal Group

> Hybrid KEM frameworks that use a KEM for the traditional component invoke the

Author: ```suggestion
Hybrid KEM frameworks that use a nominal group for the traditional component invoke the
```

DeriveKeyPair, Encap, and Decap functions of PQ KEMs, alongside analogous
functions of the nominal group.  The "encapsulation key" is the receiver's
public group element; the "ciphertext" is an ephemeral group element; and
the shared secret is the secret value resulting from an ephemeral-static
Diffie-Hellman exchange.

~~~ pseudocode
def expandDecapsKeyG(seed):
    seed_full = PRG(seed)
    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, Group_T.Nseed, seed_full)

    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
    dk_T = Group_T.RandomScalar(seed_T)
    ek_T = Group_T.Exp(Group_T.g, dk_T)

    return (ek_PQ, ek_T, dk_PQ, dk_T)

def prepareEncapG(ek_PQ, ek_T):

>     (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)

Reviewer: ```suggestion
    (ss_PQ, ct_PQ) = KEM_PQ.Encaps(ek_PQ)
```

    sk_E = Group_T.RandomScalar(random(Group_T.Nseed))
    ct_T = Group_T.Exp(Group_T.g, sk_E)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ek_T, sk_E))
    return (ss_PQ, ss_T, ct_PQ, ct_T)

> def prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T):

Reviewer: ```suggestion
def prepareDecapsG(ct_PQ, ct_T, dk_PQ, dk_T):
```

> def prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T):

Reviewer: Not sure about 'prepare', this is doing a lot of the work of `Encaps`/`Decaps`, maybe 'process'?

Author: "Prepare" in the sense of "preparing the inputs to the KDF"

    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
    return (ss_PQ, ss_T)
~~~

### Using a Traditional KEM

Hybrid KEM frameworks that use a KEM for the traditional component invoke the

> DeriveKeyPair, Encap, and Decap functions of the traditional and PQ KEMs in

Reviewer: ```suggestion
`DeriveKeyPair`, `Encaps`, and `Decaps` functions of the traditional and PQ KEMs in
```

Author: https://github.com/cfrg/draft-irtf-cfrg-hybrid-kems/issues/86

parallel.

~~~ pseudocode
def expandDecapsKeyK(seed):
    seed_full = PRG(seed)
    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, KEM_T.Nseed, seed_full)
    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
    (ek_T, dk_T) = KEM_T.DeriveKeyPair(seed_T)
    return (ek_PQ, ek_T, dk_PQ, dk_T)

def prepareEncapK(ek_PQ, ek_T):
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
    (ss_T, ct_T) = KEM_T.Encap(ek_T)
    return (ss_PQ, ss_T, ct_PQ, ct_T)

> def prepareDecapK(ct_PQ, ct_T, dk_PQ, dk_T):

Reviewer: ```suggestion
def prepareDecapsK(ct_PQ, ct_T, dk_PQ, dk_T):
```

    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
    ss_T = KEM_T.Decap(dk_T, ct_T)
    return (ss_PQ, ss_T)
~~~

### Combiners

A combiner function uses the KDF used in the hybrid KEM to combine the shared

> secrets output by the constituent algorithms with metadata.  This operation is

Author: ``` suggestion
secrets output by the component algorithms with contextual information. This operation is
```

You say "component", I say "constituent", let's pick one and not both :)  Seems like "component is the one more people like.

important to the IND-CCA security of the hybrid KEM, as well as its binding
properties.

The two combiner functions defined in this document are as follows:

~~~

> def UniversalCombiner(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label):

Author: "Universal" in the sense that it is secure for any KEM, not just a C2PRI KEM.  Maybe that's the way to phrase it: "any KEM" vs. "only C2PRI".  I will probably merge this PR to get the structure in, and then we can iterate on the naming.

https://github.com/cfrg/draft-irtf-cfrg-hybrid-kems/issues/87

    KDF(concat(ss_PQ, ss_T, ct_PQ, ct_T, ek_PQ, ek_T, label))

def C2PRICombiner(ss_PQ, ss_T, ct_T, ek_T, label):
    KDF(concat(ss_PQ, ss_T, ct_T, ek_T, label))
~~~

Note that while the names of the inputs are suggestive of the shared secret,
ciphertext, and encapsulation key outputs of a KEM, the inputs to this function
in the hybrid KEM framework are not necessarily the output of a secure KEM.  In
particular, when the framework is instantiated with a nominal group, the
"ciphertext" component is an ephemeral group element, and the "encapsulation

> key" is the group element that functions as the recipient's public key.

Reviewer: ```suggestion
in the hybrid KEM framework are not necessarily the output of a secure KEM. In
particular, when the framework is instantiated with a nominal group, the
"ciphertext" component is an ephemeral group element, and the "encapsulation
key" is the group element that functions as the recipient's public key.
```

The choice of combiner in a given instantitation determines the assumptions
under which the resulting hybrid KEM is secure.

The `UniversalCombiner` combiner explicitly hashes shared secrets, ciphertexts, and
encapsulation keys from both constituent.  This allows the resulting hybrid KEM
to be secure as long as either component is secure, with no further assumptions.

The `C2PRICombiner` combiner does not explicitly hash in the ciphertext or
encapsulation key from the PQ constituent.  The resulting hybrid KEM will only
be secure if, in addition to one constituent being secure, the PQ constituent
also satisfies the C2PRI property.

## Key Generation

All four frameworks share a common key generation function:

~~~
def GenerateKeyPair():
    seed = random(Nseed)
    return DeriveKeyPair(seed)
~~~

In some deployment environments, it is not possible to instantiate this process.

> Some implementations of constituent algorithms do not support the DeriveKeyPair

Reviewer: ```suggestion
Some implementations of component schemes do not support the `DeriveKeyPair`
```

an attacker can cause the victim to use a crafted decapsulation key. The above
hybrid KEM framework assures binding properties in the face of a LEAK attacker,
irrespective of how key generation is done. The additional provided by the
default "shared seed" key generation upgrades this to protection against a MAL
attacker.

As a result, this alternative style of key generation should only be used in
environments where implementations of component algorithms do not allow
decapsulation keys to be imported or exported.  When decapsulation keys are
moved around, they MUST always use the seed format.

## UG Framework: Universal Combiner with a Nominal Group

This framework combines a PQ KEM with a nominal group, using the universal
combiner function.  It should be used in cases where the application wants to
use a nominal group for the traditional component, and does not want to rely on
the C2PRI assumption for the PQ KEM.

~~~
def DeriveKeyPair(seed):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(seed)
    return (concat(ek_PQ, ek_T), seed)

def Encaps(ek):
    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)
    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapG(ek_PQ, ek_T)

>     ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))

Reviewer: The invocation of `*Combiner` seems to be doing less than the 'prepare*'/'process' functions, I wonder if we can just inline them (as one line each)

Author: I don't feel strongly, but I kind of like having the indirection because (a) it highlights the commonality and (b) having a name for it highlights the security assumptions.

    ct_H = concat(ct_PQ, ct_T)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(dk)
    (ss_PQ, ss_T) = prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T)
    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
    return ss_H
~~~

> ## UK Framework: Universal Combiner with a KEM

Reviewer: Repeating 'meh'-ness with 'Universal'

This framework combines a PQ KEM with a traditional KEM, using the universal
combiner function.  It should be used in cases where the application wants to
use a KEM for the traditional component, and does not want to rely on the C2PRI
assumption for the PQ KEM.

~~~
def DeriveKeyPair(seed):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyK(seed)

>     return (concat(ek_PQ, ek_T), seed)

Reviewer: Other places have the secret key first, left to right

def Encaps(ek):
    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)

>     (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapK(ek_PQ, ek_T)

Reviewer: ```suggestion
    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapsK(ek_PQ, ek_T)
```

    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))

>     ct_H = concat(ct_PQ, ct_T)

Reviewer: We do `concat()` on their own lines in some places but not others, should we try to be consistent?

    return (ss_H, ct_H)

def Decaps(dk, ct):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyK(dk)
    (ss_PQ, ss_T) = prepareDecapK(ct_PQ, ct_T, dk_PQ, dk_T)
    ss_H = UniversalCombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
    return ss_H
~~~

## CG Framework: C2PRI Combiner with a Nominal Group

This framework combines a PQ KEM with a nominal group, using the C2PRI combiner
function.  It should be used in cases where the application wants to use a
nominal group for the traditional component, and is comfortable relying on the
C2PRI assumption for the PQ KEM.

~~~
def DeriveKeyPair(seed):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(seed)
    return (concat(ek_PQ, ek_T), seed)

def Encaps(ek):
    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)

>     (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapG(ek_PQ, ek_T)

Reviewer: ```suggestion
    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapsG(ek_PQ, ek_T)
```

    ss_H = C2PRICombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
    ct_H = concat(ct_PQ, ct_T)
    return (ss_H, ct_H)

def Decaps(dk, ct):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyG(dk)

>     (ss_PQ, ss_T) = prepareDecapG(ct_PQ, ct_T, dk_PQ, dk_T)

Reviewer: ```suggestion
    (ss_PQ, ss_T) = prepareDecapsG(ct_PQ, ct_T, dk_PQ, dk_T)
```

    ss_H = C2PRICombiner(ss_PQ, ss_T, ct_T, ek_T, Label))
    return ss_H
~~~

## CG Framework: C2PRI Combiner with a KEM

This framework combines a PQ KEM with a traditional KEM, using the C2PRI
combiner function.  It should be used in cases where the application wants to
use a KEM for the traditional component, and is comfortable relying on the C2PRI
assumption for the PQ KEM.

~~~
def DeriveKeyPair(seed):
    (ek_PQ, ek_T, dk_PQ, dk_T) = expandDecapsKeyK(seed)
    return (concat(ek_PQ, ek_T), seed)

def Encaps(ek):
    (ek_PQ, ek_T) = split(KEM_PQ.Nek, Group_T.Nelem, ek)

>     (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapK(ek_PQ, ek_T)

Reviewer: ```suggestion
    (ss_PQ, ss_T, ct_PQ, ct_T) = prepareEncapsK(ek_PQ, ek_T)
```

def prepareDecapsG(ct_PQ, ct_T, dk_PQ, dk_T):
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
    ss_T = Group_T.ElementToSharedSecret(Group_T.Exp(ct_T, dk_T))
    return (ss_PQ, ss_T)
~~~

### Using a Traditional KEM

Hybrid KEM frameworks that use a KEM for the traditional component invoke the
DeriveKeyPair, Encap, and Decap functions of the traditional and PQ KEMs in
parallel.

~~~ pseudocode
def expandDecapsKeyK(seed):
    seed_full = PRG(seed)
    (seed_PQ, seed_T) = split(KEM_PQ.Nseed, KEM_T.Nseed, seed_full)
    (ek_PQ, dk_PQ) = KEM_PQ.DeriveKeyPair(seed_PQ)
    (ek_T, dk_T) = KEM_T.DeriveKeyPair(seed_T)
    return (ek_PQ, ek_T, dk_PQ, dk_T)

def prepareEncapK(ek_PQ, ek_T):
    (ss_PQ, ct_PQ) = KEM_PQ.Encap(ek_PQ)
    (ss_T, ct_T) = KEM_T.Encap(ek_T)
    return (ss_PQ, ss_T, ct_PQ, ct_T)

def prepareDecapK(ct_PQ, ct_T, dk_PQ, dk_T):
    ss_PQ = KEM_PQ.Decap(dk_PQ, ct_PQ)
    ss_T = KEM_T.Decap(dk_T, ct_T)
    return (ss_PQ, ss_T)
~~~

### Combiners

A combiner function uses the KDF used in the hybrid KEM to combine the shared
secrets output by the constituent algorithms with metadata.  This operation is
important to the IND-CCA security of the hybrid KEM, as well as its binding

> properties.

Author: ```suggestion
secrets output by the component algorithms with contextual information.
```
